compliance/collection/main.go

```diff
@@ -105,21 +108,13 @@ func startAuditLogCollection(ctx context.Context, client sensor.ComplianceServic
 	return auditReader
 }
 
-func manageStream(ctx context.Context, cli sensor.ComplianceServiceClient, sig *concurrency.Signal) {
+func manageStream(ctx context.Context, client sensor.ComplianceService_CommunicateClient, config *sensor.MsgToCompliance_ScrapeConfig, sig *concurrency.Signal) {
 	for {
 		select {
 		case <-ctx.Done():
 			sig.Signal()
 			return
 		default:
-			client, config, err := initializeStream(ctx, cli)
-			if err != nil {
-				if ctx.Err() != nil {
-					// continue and the <-ctx.Done() path should be taken next iteration
-					continue
-				}
-				log.Fatalf("error initializing stream to sensor: %v", err)
-			}
 			if err := runRecv(ctx, client, config); err != nil {
 				log.Errorf("error running recv: %v", err)
 			}
```

Comment:
@vikin91 @Maddosaurus 
Unless I read the code incorrectly, the `manageStream` function does not function the way it supposed to.
I think, its idea should be to check whether the connection got closed (`ctx.Done()`) and in this case shut down the process (`sig.Signal()`), or (if the connection is still alive) receive a message from Sensor and respond back (`runRecv()`). The problem is that `runRecv()` is a blocking function that runs an infinite loop of receive-reply logic. Once the code gets to it, it will be running forever until an error happens or unsupported message arrives.
When the code is in `runRecv()`, the shutdown logic of `manageStream` won't be executed normally.
Am I right or am I missing something?

Reply:
I definitely want to look at sensor and see how it is done there.

For compliance, I see that `client.Recv()` is being called on a client that was created with the context `client, err := cli.Communicate(ctx)`, so I hope that canceling the context would shutdown the client and release the lock.
(I used word "hope" as I have not verified this and do not know the internals of grpc stream).

I will do some more reading on this.

Reply:
So sensor utilizes two clever objects to manage the stream to central: a pointer `*grpcUtil.LazyClientConn` holding a lazy connection, and `centralclient.CentralConnectionFactory` that sets, monitors the connection and restarts it when failed. All communication attempts are blocking and waiting when the connection is down (no message is lost). The operations resume when the connection is back up.

We could apply this to compliance-sensor as well, but this may deserve a separate refactoring PR. Would it be okay to continue in the current shape and improve this in a separate PR @msugakov ?

Reply:
The code has changed now a lot. Issues that were before are now gone but new things still don't look 100% self-explanatory to me. I think we (I?) need to confidently understand well what we do. I would not classify "understanding" as "improvement" or "optimization". "Understanding" is our core skill.

Reply:
Yes, sorry for adding so many changes after you started the review. I think they were needed to address some obvious drawbacks regarding sending without being sure that the connection is up.

While there are still some bits about the GRPC stream that I would like to understand better, I feel pretty confident that the code handling sending and receiving in this component follows the same principles as in other parts of our code.